### PR TITLE
core/txpool, eth: Introduce delay between broadcasting/announcing consecutive local transactions from the same account

### PR DESCRIPTION
**Pull request by jwasinger**
## Problem Explanation

The goal of this change is to implement a basic heuristic solution to improve the flow of nonce-linked of transactions through the Ethereum network. Take as example, a situation the user wants to send a list of nonce-contiguous locally-originating transactions.

Currently, these will be immediately sent nonce-ordered to a square-root subset of the peer set and announced to the rest. As the sequence of transactions propagate throughout the network, a situation may occur where a node, being unable to know the nonce-ordering of the set of announced transaction hashes (and/or having a non-matching chain state compared to its peers), receives a subset (whether by request/receive, direct-broadcast or a combination of both) that is not executable.

In Geth, the transaction pool only reserves a certain number of slots per-account to queue non-executable transactions before dropping any subsequently received non-executable transactions from the account. The affected account cannot become cleared unless the node is restarted or head chain state is adopted which includes more transactions from the account (bumping the nonce to the point where transactions stuck in the queue are either removed, or become executable).

The changes here aim to reduce the likelihood of this situation by increasing the chance that a given node will receive an executable portion of a sequence of nonce-linked transactions.  This is because the executable portion becomes visible earlier on the network.

## Solution Proposed/Implemented in this PR

This PR changes (non-blob) transaction announcement/broadcast to institute a delay of 400ms between announcing/broadcasting transactions from the same local account.

Previously, the implementation could send transactions/hashes to peers as they arrived from the transaction pool, requiring no outside-of-txpool tracking beyond remembering which peers had been sent/announced which transactions.

However, with the introduction of a delay between becoming executable and being broadcasted/announced, local transactions must be potentially queued and kept.  In my implementation, executable transactions are forwarded to a '`localsTxHandler`' after they move through the txpool (become executable). The `localsTxsHandler` adds each tx to a nonce-ordered list associated with the sender account.  Lower nonce transactions are removed from the queues if their nonce falls below that of their sender account in the head state of the chain.  The queues are not constrained from growing arbitrarily large with higher nonce transactions because there is already precedence for this with the txpool tracking arbitrarily-many executable transactions per sender.

For each peer, we track the highest nonce and timestamp that we last sent a transaction/hash to.  Upon receiving new local transactions from the txpool or timing out 400ms since we last broadcasted/announced to a peer, we gather and send all local transactions/hashes that can be sent to each peer.

#### Testing

I have created unit tests which are a bit of a hack and make use of utilities in the existing `eth` devp2p test suite.  However, they require the ability to send signed transactions to a node so it doesn't make sense for them to be a part of this suite.

The options I can think of are:
* keep them as unit tests in the Geth codebase.
* create a new black-box Hive test suite where a node is required to expose `eth_sendSignedTransaction`, the engine API (to have a mock-CL drive block production), other rpc methods necessary to poll the client for created blocks.

#### User "friction"

A delay of 400ms is chosen because it is slightly less than the tx fetcher's timeout to retrieve an announced transaction.  This will inhibit use-cases where users want to spam out transactions in quick succession
(such as in private networks).  It's not clear to me what the ideal solution to this is right now.

### CODE REVIEW

**Comment by jwasinger** [comment]
Looking at the incoming 4844 changes, I'm starting to think that I will have to significantly rethink the approach here.  Tabling any further development on this until I have a clear understanding of the "broader picture" of incoming changes.

File: eth/local_tx_handler.go [hunk 1]
```diff
@@ -0,0 +1,497 @@
...
+
+	return res
+}
+
+// TODO: noting an edge-case/idea/question here.  If we have already broadcasted/announced a set of transactions to a peer and we receive one or more replacement transactions with the same nonce
+// should we broadcast/announce them all immediately?  seems like a no-brainer.
+
+// mergeNonceOrderedTxs combines curTxs and newTxs together.  each list is nonce-contiguous, nonce-unique ordered ascending by nonce.  If the nonce of a transaction in newTxs matches a transaction in curTxs, only the transaction from newTxs is included in the result.
+// The resulting list is nonce-contiguous, nonce-unique and ordered ascending by nonce.
+func mergeNonceOrderedTxs(curTxs []*types.Transaction, newTxs []*types.Transaction) []*types.Transaction {
```
Comment (holiman) [inline eth/local_tx_handler.go:364]: I got stuck here a bit, this seemed like the kind of thing that might be nicer to do with generics? Here's an untested proof of concept

```
func mergeNonceOrderedTxs(curTxs []*types.Transaction, newTxs []*types.Transaction) []*types.Transaction {
	cmpFn := func(a, b *types.Transaction) {
		if a.Nonce() < b.Nonce() {
			return -1
		}
		if a.Nonce > b.Nonce() {
			return 1
		}
		return 0
	}
	return mergeNonceOrderedGeneric(curTxs, newTxs, cmpFn)
}

func mergeNonceOrderedGeneric[T any](listA []T, listB []T, cmpFn func(a, b T) int) []T {
	// insert new txs into the sender's queue, keeping the resulting array nonce-ordered and
	// replacing pre-existing txs if there is a tx with same nonce
	var (
		res  []T
		idxA = 0
		idxB = 0
	)
	for {
		if idxA >= len(listA) || idxB >= len(listB) {
			break
		}
		txA, txB := listA[idxA], listB[idxB]
		switch cmp(txA, txB) {
		case -1:
			res = append(res, txA)
			idxA++
		case 0:
			idxA++
			fallthrough
		default:
			res = append(res, txB])
			idxB++
		}
	}
	// Merge is done, at least one of them is empty, just add whatever remains
	if idxA < len(listA) {
		res = append(res, listA[idxA:]...)
	}
	if idxB < len(listB) {
		res = append(res, listB[idxB:]...)
	}
	return res
}
```

I have a feeling it is cleaner that way: the sorting doesn't know about tx characteristics, the algo is fully generic.

File: eth/local_tx_handler.go [hunk 2]
```diff
@@ -0,0 +1,497 @@
...
+
+	curTxs, ok := l.txQueues[sender]
+	if !ok {
+		l.txQueues[sender] = newTxs
+		l.updateLowestUnsentNonce(sender, newTxs[0].Nonce())
+		l.updateLowestUnannouncedNonce(sender, newTxs[0].Nonce())
+		return
+	}
+
+	newTxsProceedCurrent := newTxs[0].Nonce() > l.txQueues[sender][len(l.txQueues[sender])-1].Nonce()
```
Comment (holiman) [inline eth/local_tx_handler.go:429]:
```code
	newTxsProceedCurrent := newTxs[0].Nonce() > l.txQueues[sender][len(l.txQueues[sender])-1].Nonce()
```
Wow, what a mouthful: `l.txQueues[sender][len(l.txQueues[sender])-1].Nonce()`. 

```suggestion
	newTxsProceedCurrent := newTxs[0].Nonce() > curTxs[len(curTxs) -1].Nonce()
```

**Review by holiman — COMMENTED** [review]
Want to add what I said in the triage. Over all, my feeling is that this PR strives to be "too perfect". It tries to model the state of the world around it (the state of the peers) too carefully, keeping track of what is sent to whom and when. 

All in all, it becomes a pretty complex machinery, with, I'm sure, it's own cornercases and flaws that potentially backfire at some point, when these assumptions do not match reality. 

So, right now, we don't do anything to try and solve the problem with "a massive amount of txs that we send out en masse". This PR does it, IMO, too extensively and thorougly. 

I'm kind of looking for a middle ground, where we do a little bit of extra work. Some simplistic local delay -- maybe a standalone delay-line between the RPC submission endpoint and the transaction pool entry would suffice? I don't know, sorry.

**Comment by jwasinger** [comment]
> maybe a standalone delay-line between the RPC submission endpoint and the transaction pool entry would suffice?

👍 This feels like a good way to achieve what this PR attempts to with less complexity.

File: eth/local_tx_handler.go (hunk 1 above)
Comment (fjl) [reply eth/local_tx_handler.go:364]: No need to make code generic unless there is an actual need to reuse it with multiple types.

**Pull request closed without merging**